Source/NSAttributedString.m

```diff
@@ -1095,20 +1096,43 @@ - (NSFileWrapper *) fileWrapperFromRange: (NSRange)range
 - (NSInteger) itemNumberInTextList: (NSTextList *)list
                            atIndex: (NSUInteger)location
 {
-  NSParagraphStyle *style = [self attribute: NSParagraphStyleAttributeName
-                                  atIndex: location
-                                  effectiveRange: NULL];
-  if (style != nil)
+  NSRange listRange = [self rangeOfTextList: list atIndex: location];
+  if (listRange.location == NSNotFound)
     {
-      NSArray *textLists = [style textLists];
+      return 0;
+    }
 
-      if (textLists != nil)
+  NSRange subRange = NSMakeRange(listRange.location, location-listRange.location+1);
+  unichar buffer[subRange.length];
+  
+  [[self string] getCharacters: buffer range: subRange];
+  
+  NSCharacterSet *newlineCharacterSet = [NSCharacterSet newlineCharacterSet];
+  
+  NSUInteger itemNumber = 1;
+  NSUInteger index;
+  for (index=1; index<subRange.length; index++)
+    {
+      if ([newlineCharacterSet characterIsMember: buffer[index-1]])
         {
-          return [textLists indexOfObject: list];
+          NSParagraphStyle *style = [self attribute: NSParagraphStyleAttributeName
+            atIndex: subRange.location + index
+            effectiveRange: NULL];
+          
+          NSArray *textLists = [style textLists];
+          if ([list isEqual: [textLists lastObject]])
+            {
+              itemNumber++;
+            }
+          
+          if (buffer[index-1] == '\r' && buffer[index] == '\n')
```

Comment:
This and the use of `NSCharacterSet newlineCharacterSet` may be overkill, would we use '\n' internally or do we need to support Windows CR+LF combinations?